app/views/articles/_single_story.html.erb

```diff
@@ -35,9 +35,16 @@
         </div>
         <div>
           <p>
-            <a href="/<%= story.cached_user.username %>" class="crayons-story__secondary fw-medium">
+            <a href="/<%= story.cached_user.username %>" class="crayons-story__secondary fw-medium m:hidden">
               <%= story.cached_user.name %>
             </a>
+            <div class="profile-preview-card relative mb-4 s:mb-0 fw-medium hidden m:inline-block">
+              <button id="story-author-preview-trigger-<%= story.id %>" aria-controls="story-author-preview-content-<%= story.id %>" class="profile-preview-card__trigger fs-s px-0 crayons-btn crayons-btn--ghost p-0" aria-label="<%= story.cached_user.username %> profile details">
+                <%= story.cached_user.name %>
+              </button>
+              <%= render "/shared/profile_preview_card", actor: story.user, id: "story-author-preview-content-#{story.id}" %>
```

Comment:
Unfortunately this partial is rendered by so many different pages that it's complicated to understand which `story` is referring to:

```shell
> rg "single_story"
app/views/collections/show.html.erb
18:          <%= render "articles/single_story", story: article.decorate, featured: article.main_image.present? %>

app/views/users/_main_feed.html.erb
10:      <%= render partial: "articles/single_story", collection: @pinned_stories, as: :story, locals: { featured: false } %>
15:  <%= render "articles/single_story", story: @stories.first, featured: false %>
19:  <%= render partial: "articles/single_story", collection: @stories[1, @stories.size], as: :story, locals: { featured: false } %>

app/views/stories/_main_stories_feed.html.erb
2:  <%= render partial: "articles/single_story", locals: { story: @pinned_article, pinned: true, featured: @pinned_article.id == @featured_story.id } %>
6:  <%= render partial: "articles/single_story", locals: { story: @featured_story, featured: true, pinned: false } %>
20:      <%= render partial: "articles/single_story", locals: { story: story, featured: false, pinned: false } %>

app/views/stories/tagged_articles/_main_feed.html.erb
1:<%= render partial: "articles/single_story", collection: @stories, as: :story, locals: { featured: false } %>

app/views/organizations/_main_feed.html.erb
2:<%= render partial: "articles/single_story", collection: @stories, as: :story, locals: { featured: false } %>
```

That said, since it's about the feed, it means that we need to go through all the places these files above are rendered from, find the corresponding controller action and load the `user` object, by adding `.includes([:user])` to the query.

I think `app/views/collections/show.html.erb` is fine since it's a `show` action (the error is present because its trying to render a list of objects, which incurs into a N + 1 query - each iteration loads the user object from the DB).

A place to start is `app/views/users/_main_feed.html.erb` which is rendered by `app/views/users/show.html.erb` which is rendered by https://github.com/forem/forem/blob/0ccb7608d0326d8320aea2303f7b280b87164458/app/controllers/stories_controller.rb#L213 which is basically the user profile page IIRC.

At some point in that method, the method `assign_user_stories` is called, which defines `@stories` and `@pinned_stories`:

https://github.com/forem/forem/blob/0ccb7608d0326d8320aea2303f7b280b87164458/app/controllers/stories_controller.rb#L311-L319

The statement `.includes([:user])` isn't present, which is why the test fails.

That said: which attributes of the `User` do you need, because IIRC the reason `_single_story` is using `cached_user` is for performance reasons and to avoid loading the entire user object in memory, maybe there's an alternative solution here where we add the required attributes to `.cached_user`...

Reply:
Thanks for all this detail - it's so helpful!

Yes, I was wondering if maybe the right solution would be to add to the cached user. As things stand at the moment, the attributes needed are:

- all the metadata that we show on the card (location, work, etc)
- the `user_colors` method which relies on `user.decorate.enriched_colors` and all that touches further down the line 😩 

It might be worth saying though that this is only affecting the **logged out** feed. Because that feed is initially generated server-side, I added the preview card in its entirety to the `html.erb` template. An alternative solution is that I handle the logged-out feed the same way as the logged in/paged in feed - i.e. show a minimal preview card (which I believe would only need the cached user details), and then fetch the remaining data view network call on hover.

I think that probably makes the most sense. I was trying to avoid unnecessary data fetches from the frontend, but stuffing loads of extra stuff in the cached user doesn't seem right for a feature a person might never use...

Reply:
I like this last solution!

> I think that probably makes the most sense. I was trying to avoid unnecessary data fetches from the frontend, but stuffing loads of extra stuff in the cached user doesn't seem right for a feature a person might never use...

Also: the `cached_user` would need to be updated continuously and some of the data is not in the `users` table anymore, so it could be tricky

Reply:
Cool - sounds like a plan then! 😎  Thanks for your help!